pwiz_tools/Shared/CommonMsData/RemoteApi/WatersConnect/WatersConnectSessionAcquisitionMethod.cs

```diff
@@ -0,0 +1,242 @@
+﻿/*
+ * Original author: Matt Chambers <matt.chambers42 .at. gmail.com>
+ *
+ * Copyright 2024 University of Washington - Seattle, WA
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using Newtonsoft.Json.Linq;
+using pwiz.Common.Collections;
+using pwiz.Common.SystemUtil;
+
+namespace pwiz.CommonMsData.RemoteApi.WatersConnect
+{
+    public class WatersConnectSessionAcquisitionMethod : WatersConnectSession
+    {
+        public static readonly string GET_METHODS_ENDPOINT = @"/waters_connect/v2.0/published-methods?methodTypeIds=17C19CE93BBA488A975B1E14AAAA0B9B";
+        public static readonly string UPLOAD_METHOD_ENDPOINT = @"/waters_connect/method-develop/v1.0/tandem/create-mrm-methods";
+
+        public WatersConnectSessionAcquisitionMethod(WatersConnectAccount account) : base(account)
+        {
+        }
+
+        public override bool AsyncFetchContents(RemoteUrl remoteUrl, out RemoteServerException remoteException)
+        {
+            if (!(remoteUrl is WatersConnectUrl wcUrl))
+                throw new ArgumentException();
+
+            if (wcUrl.Type == WatersConnectUrl.ItemType.folder_with_methods)
+            {
+                RemoteServerException ex1, ex2 = null;
+                var res1 = AsyncFetch(GetRootContentsUrl(), GetFolders, out ex1);
+
+                ImmutableList<WatersConnectFolderObject> folders;
+                bool canReadMethods = false;
+                if (TryGetData(GetRootContentsUrl(), out folders))
+                {
+                    var currentFolder = folders.FirstOrDefault(f => f.Path.Equals(WebUtility.UrlDecode(wcUrl.EncodedPath)));
+                    canReadMethods = currentFolder?.CanRead ?? false;
+                }
+                if (canReadMethods)
+                {
+                    var res2 = AsyncFetch(GetAcquisitionMethodsUrl(wcUrl), GetAcquisitionMethods, out ex2);
+                    remoteException = ex1 ?? ex2;
+                    return res1 && res2;
+                }
+
+                remoteException = ex1;
+                return res1;
+            }
+
+            throw new Exception("Url type mismatch: " + wcUrl.Type);
```

Comment:
Not sure what this means or whether it could be a user-facing error. If so, resx. If not, Assume the type at the top and the if block wouldn't be necessary.